tests/Doctrine/Tests/ORM/Mapping/ClassMetadataFactoryTest.php

```diff
@@ -451,6 +454,70 @@ public function testInheritsIdGeneratorMappingFromEmbeddable()
 
         $this->assertTrue($userMetadata->isIdGeneratorIdentity());
     }
+
+    public function testPropertyOrderNoCache()
+    {
+        $cmf = $this->createPropertyOrderFactory();
+
+        /** @var ClassMetadataInfo $adminMetadata */
+        $adminMetadata = $cmf->getMetadataFor(PropertyOrder\Admin::class);
+        $this->assertSame(
+            ['id', 'name', 'group', 'email'],
+            array_keys($adminMetadata->reflFields)
+        );
+
+        /** @var ClassMetadataInfo $userMetadata */
+        $userMetadata = $cmf->getMetadataFor(PropertyOrder\User::class);
+        $this->assertSame(
+            ['id', 'name', 'group'],
+            array_keys($userMetadata->reflFields)
+        );
+    }
+
+    public function testPropertyOrderWithCache()
+    {
+        $noCacheCmf = $this->createPropertyOrderFactory();
+        $cache      = new ArrayCache();
+
+        $addCacheMetadata = function (string $className) use ($cache, $noCacheCmf) {
+            /** @var ClassMetadataInfo $metadata */
+            $metadata             = $noCacheCmf->getMetadataFor($className);
+            $metadata->reflFields = [];
+
+            $cacheSalt = '$CLASSMETADATA'; // \Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory::$cacheSalt
+
+            $cache->save($className . $cacheSalt, $metadata);
+        };
+        $addCacheMetadata(PropertyOrder\Admin::class);
+        $addCacheMetadata(PropertyOrder\User::class);
+
+        $cmf = $this->createPropertyOrderFactory();
+        $cmf->setCacheDriver($cache);
+
+        /** @var ClassMetadataInfo $adminMetadata */
+        $adminMetadata = $cmf->getMetadataFor(PropertyOrder\Admin::class);
+        $this->assertSame(
+            ['id', 'name', 'group', 'email'],
```

Comment:
Yes, order of the fields is the problem.

At least persisting new entities relies on this order. The field order is potentially relevant at least in the following places:

- [`BasicEntityPersister:1428`](https://github.com/doctrine/doctrine2/blob/388afb46d0cb3ed0c51332e8df0de9e942c2690b/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php#L1428)
- [`JoinedSubclassPersister:534`](https://github.com/doctrine/doctrine2/blob/388afb46d0cb3ed0c51332e8df0de9e942c2690b/lib/Doctrine/ORM/Persisters/Entity/JoinedSubclassPersister.php#L534)
- [`UnitOfWork:580`](https://github.com/doctrine/doctrine2/blob/388afb46d0cb3ed0c51332e8df0de9e942c2690b/lib/Doctrine/ORM/UnitOfWork.php#L580)
- [`UnitOfWork:940`](https://github.com/doctrine/doctrine2/blob/388afb46d0cb3ed0c51332e8df0de9e942c2690b/lib/Doctrine/ORM/UnitOfWork.php#L940)

The field order of the derived collections like `$columns` and `$actualData` is relevant to the generated queries.